src/uct/ze/copy/ze_copy_ep.c

```diff
@@ -0,0 +1,165 @@
+{
+    uct_ze_copy_iface_t *iface = ucs_derived_of(params->iface,
+                                                uct_ze_copy_iface_t);
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
+
+    return UCS_OK;
+}
+
+static UCS_CLASS_CLEANUP_FUNC(uct_ze_copy_ep_t)
+{
+}
+
+UCS_CLASS_DEFINE(uct_ze_copy_ep_t, uct_base_ep_t)
+UCS_CLASS_DEFINE_NEW_FUNC(uct_ze_copy_ep_t, uct_ep_t, const uct_ep_params_t*);
+UCS_CLASS_DEFINE_DELETE_FUNC(uct_ze_copy_ep_t, uct_ep_t);
+
+#define uct_ze_copy_trace_data(_remote_addr, _rkey, _fmt, ...) \
+    ucs_trace_data(_fmt " to %" PRIx64 "(%+ld)", ##__VA_ARGS__, \
+                   (_remote_addr), (_rkey))
+
+ucs_status_t uct_ze_copy_ep_zcopy(uct_ep_h tl_ep, uint64_t remote_addr,
+                                  const uct_iov_t *iov, uct_rkey_t rkey,
+                                  int is_put)
+{
+    size_t size                = uct_iov_get_length(iov);
+    uct_ze_copy_iface_t *iface = ucs_derived_of(tl_ep->iface,
+                                                uct_ze_copy_iface_t);
+    ze_result_t ret;
+    void *src, *dst;
+
+    ucs_trace("remote addr %p rkey %p size %zu", (void*)remote_addr,
+              (void*)rkey, size);
+
+    if (is_put) { /* Host-to-Device */
+        src = iov->buffer;
+        dst = (void*)remote_addr;
+    } else { /* Device-to-Host */
+        src = (void*)remote_addr;
+        dst = iov->buffer;
+    }
+
+    ret = zeCommandListAppendMemoryCopy(iface->ze_cmdl, dst, src, size, NULL, 0,
+                                        NULL);
+    if (ret != ZE_RESULT_SUCCESS) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ret = zeCommandListClose(iface->ze_cmdl);
+    if (ret != ZE_RESULT_SUCCESS) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ret = zeCommandQueueExecuteCommandLists(iface->ze_cmdq, 1, &iface->ze_cmdl,
+                                            NULL);
+    if (ret != ZE_RESULT_SUCCESS) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ret = zeCommandQueueSynchronize(iface->ze_cmdq, UINT32_MAX);
+    if (ret != ZE_RESULT_SUCCESS) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ret = zeCommandListReset(iface->ze_cmdl);
+    if (ret != ZE_RESULT_SUCCESS) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_trace("ze memory copy from src %p to dst %p, len %ld", src, dst, size);
+
+    return UCS_OK;
+}
+
+ucs_status_t uct_ze_copy_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
+                                      size_t iovcnt, uint64_t remote_addr,
+                                      uct_rkey_t rkey, uct_completion_t *comp)
+{
+    ucs_status_t status;
+
+    status = uct_ze_copy_ep_zcopy(tl_ep, remote_addr, iov, rkey, 0);
+
+    UCT_TL_EP_STAT_OP(ucs_derived_of(tl_ep, uct_base_ep_t), GET, ZCOPY,
+                      uct_iov_total_length(iov, iovcnt));
+    uct_ze_copy_trace_data(remote_addr, rkey, "GET_ZCOPY [length %zu]",
+                           uct_iov_total_length(iov, iovcnt));
```

Comment:
I would keep it here so that to distinguish PUT_ZCOPY/GET_ZCOPY/PUT_ZHORT/GET_SHORT. In addition, I am going to use `ucs_trace_data` directly here and remove that extra macro definition.